x-pack/plugins/lens/public/indexpattern_datasource/field_list.tsx

```diff
@@ -0,0 +1,200 @@
+  fieldGroups,
+  existenceFetchFailed,
+  fieldProps,
+  hasSyncedExistingFields,
+  filter,
+  currentIndexPatternId,
+}: {
+  exists: (field: IndexPatternField) => boolean;
+  fieldGroups: FieldGroups;
+  fieldProps: FieldItemSharedProps;
+  hasSyncedExistingFields: boolean;
+  existenceFetchFailed?: boolean;
+  filter: {
+    nameFilter: string;
+    typeFilter: string[];
+  };
+  currentIndexPatternId: string;
+}) {
+  const [pageSize, setPageSize] = useState(PAGINATION_SIZE);
+  const [scrollContainer, setScrollContainer] = useState<Element | undefined>(undefined);
+  const [accordionState, setAccordionState] = useState<Partial<Record<string, boolean>>>(() =>
+    Object.fromEntries(
+      Object.entries(fieldGroups)
+        .filter(([, { showInAccordion }]) => showInAccordion)
+        .map(([key, { isInitiallyOpen }]) => [key, isInitiallyOpen])
+    )
+  );
+
+  const isAffectedByFieldFilter = !!(filter.typeFilter.length || filter.nameFilter.length);
+
+  useEffect(() => {
+    // Reset the scroll if we have made material changes to the field list
+    if (scrollContainer) {
+      scrollContainer.scrollTop = 0;
+      setPageSize(PAGINATION_SIZE);
+    }
+  }, [filter.nameFilter, filter.typeFilter, currentIndexPatternId, scrollContainer]);
+  // const currentIndexPattern = indexPatterns[currentIndexPatternId];
+  // const allFields = currentIndexPattern.fields;
+  // const clearLocalState = () => setLocalState((s) => ({ ...s, nameFilter: '', typeFilter: [] }));
+  // const hasSyncedExistingFields = existingFields[currentIndexPattern.title];
+  // const availableFieldTypes = uniq(allFields.map(({ type }) => type)).filter(
+  //   (type) => type in fieldTypeNames
+  // );
+
+  const lazyScroll = useCallback(() => {
+    if (scrollContainer) {
+      const nearBottom =
+        scrollContainer.scrollTop + scrollContainer.clientHeight >
+        scrollContainer.scrollHeight * 0.9;
+      if (nearBottom) {
+        setPageSize(
+          Math.max(
+            PAGINATION_SIZE,
+            Math.min(
+              pageSize + PAGINATION_SIZE * 0.5,
+              getDisplayedFieldsLength(fieldGroups, accordionState)
+            )
+          )
+        );
+      }
+    }
+  }, [scrollContainer, pageSize, setPageSize, fieldGroups, accordionState]);
+
+  const paginatedFields = useMemo(() => {
+    let remainingItems = pageSize;
+    return Object.fromEntries(
+      Object.entries(fieldGroups)
+        .filter(([, { showInAccordion }]) => showInAccordion)
+        .map(([key, fieldGroup]) => {
+          if (!accordionState[key] || remainingItems <= 0) {
+            return [key, []];
+          }
+          const slicedFieldList = fieldGroup.fields.slice(0, remainingItems);
+          remainingItems = remainingItems - slicedFieldList.length;
+          return [key, slicedFieldList];
+        })
+    );
+  }, [pageSize, fieldGroups, accordionState]);
+
+  return (
+    <div
+      className="lnsInnerIndexPatternDataPanel__listWrapper"
```

Comment:
Don't forget now to update class names to align with the component name and move the SASS to be imported directly by the JS component.